diff -urNp coreutils-8.4-orig/src/tail.c coreutils-8.4/src/tail.c
--- coreutils-8.4-orig/src/tail.c	2014-05-05 13:56:40.117154756 +0200
+++ coreutils-8.4/src/tail.c	2014-05-05 13:59:43.311090465 +0200
@@ -1230,6 +1230,24 @@ any_remote_file (const struct File_spec 
   return false;
 }
 
+/* Return true if any of the N_FILES files in F is a symlink.
+   Note we don't worry about the edge case where "-" exists,
+   since that will have the same consequences for inotify,
+   which is the only context this function is currently used.  */
+
+static bool
+any_symlinks (const struct File_spec *f, size_t n_files)
+{
+  size_t i;
+
+  struct stat st;
+  for (i = 0; i < n_files; i++)
+    if (lstat (f[i].name, &st) == 0 && S_ISLNK (st.st_mode))
+      return true;
+  return false;
+}
+
+
 /* Return true if any of the N_FILES files in F represents
    stdin and is tailable.  */
 
@@ -2141,7 +2159,10 @@ main (int argc, char **argv)
          Note if there is a change to the original file then we'll
          recheck it and follow the new file, or ignore it if the
          file has changed to being remote.  */
-      if (tailable_stdin (F, n_files) || any_remote_file (F, n_files))
+      if (!disable_inotify && ( tailable_stdin (F, n_files)
+                              || any_remote_file (F, n_files)
+                              || any_symlinks (F, n_files)
+                              || (!ok && follow_mode == Follow_descriptor)))
         disable_inotify = true;
 
       if (!disable_inotify)
