--- coreutils-8.4/src/df.c	2014-07-16 10:33:44.606556329 -0400
+++ df.c	2014-07-16 10:27:42.680557241 -0400
@@ -793,9 +793,22 @@ get_disk (char const *disk)
   struct mount_entry const *me;
   struct mount_entry const *best_match = NULL;
 
+  char *resolved = canonicalize_file_name (disk);
+  if (resolved && resolved[0] == '/')
+    disk = resolved;
+
   for (me = mount_list; me; me = me->me_next)
-    if (STREQ (disk, me->me_devname))
-      best_match = me;
+    {
+      char *me_resolved = canonicalize_file_name (me->me_devname);
+      if (me_resolved)
+        {
+          if (STREQ (disk, me_resolved))
+            best_match = me;
+          free(me_resolved);
+        }
+    }
+
+  free(resolved);
 
   if (best_match)
     {
