diff -urNp coreutils-8.4-orig/src/chcon.c coreutils-8.4/src/chcon.c
--- coreutils-8.4-orig/src/chcon.c	2014-05-05 14:43:19.890433161 +0200
+++ coreutils-8.4/src/chcon.c	2014-05-05 14:45:28.954926163 +0200
@@ -359,6 +359,8 @@ Usage: %s [OPTION]... CONTEXT FILE...\n\
 Change the SELinux security context of each FILE to CONTEXT.\n\
 With --reference, change the security context of each FILE to that of RFILE.\n\
 \n\
+      --dereference      affect the referent of each symbolic link (this is\n\
+                         the default), rather than the symbolic link itself\n\
   -h, --no-dereference   affect symbolic links instead of any referenced file\n\
 "), stdout);
       fputs (_("\
