diff -urNp coreutils-8.4-orig/src/group-list.c coreutils-8.4/src/group-list.c
--- coreutils-8.4-orig/src/group-list.c	2010-01-03 18:06:20.000000000 +0100
+++ coreutils-8.4/src/group-list.c	2013-07-31 16:23:20.716228021 +0200
@@ -38,11 +38,14 @@ print_group_list (const char *username,
                   bool use_names)
 {
   bool ok = true;
-  struct passwd *pwd;
+  struct passwd *pwd = NULL;
 
-  pwd = getpwuid (ruid);
-  if (pwd == NULL)
-    ok = false;
+  if (username)
+    {
+      pwd = getpwuid (ruid);
+      if (pwd == NULL)
+      ok = false;
+    }
 
   if (!print_group (rgid, use_names))
     ok = false;
@@ -58,9 +61,8 @@ print_group_list (const char *username,
     gid_t *groups;
     int i;
 
-    int n_groups = xgetgroups (username, (pwd ? pwd->pw_gid : (gid_t) -1),
-                               &groups);
-    if (n_groups < 0)
+  int n_groups = xgetgroups (username, (pwd ? pwd->pw_gid : egid), &groups);
+  if (n_groups < 0)
       {
         if (username)
           {
